File: packages/py/tracing/src/uselemma_tracing/turn.py
from __future__ import annotations

import json
import uuid
from datetime import datetime
from typing import Any

from .client import Lemma, TraceContext, TraceHandle, _datetime_or_now, _iso, _now

TURN_CONTEXT_VERSION = 1
TURN_JOURNAL_VERSION = 1

_RECORD_KEYS = (
    "id",
    "parentId",
    "name",
    "type",
    "input",
    "output",
    "metadata",
    "attributes",
    "startedAt",
    "endedAt",
    "durationMs",
    "status",
    "error",
    "model",
    "toolName",
    "usage",
    "userFacingMessage",
    "llmProvider",
    "llmModelName",
    "llmInputMessages",
    "llmOutputMessages",
    "llmInvocationParameters",
    "llmTools",
)


def _compact(payload: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in payload.items() if value is not None}


def parse_turn_context_token(value: str | dict[str, Any]) -> dict[str, Any]:
    token = json.loads(value) if isinstance(value, str) else value
    if (
        not isinstance(token, dict)
        or token.get("version") != TURN_CONTEXT_VERSION
        or not isinstance(token.get("traceId"), str)
        or not token["traceId"]
        or not isinstance(token.get("startedAt"), str)
        or not token["startedAt"]
    ):
        raise ValueError("uselemma-tracing: invalid turn context token")
    return token


def _parse_journal(value: Any) -> tuple[dict[str, Any] | None, list[dict[str, Any]]]:
    parsed = json.loads(value) if isinstance(value, str) else value
    if isinstance(parsed, list):
        return None, parsed
    if isinstance(parsed, dict) and "op" in parsed:
        return None, [parsed]
    if isinstance(parsed, dict) and isinstance(parsed.get("records"), list):
        if (
            parsed.get("version") is not None
            and parsed.get("version") != TURN_JOURNAL_VERSION
        ):
            raise ValueError(
                f"uselemma-tracing: unsupported turn journal version {parsed.get('version')}"
            )
        token = parsed.get("token")
        return token if isinstance(token, dict) else None, parsed["records"]
    raise ValueError("uselemma-tracing: invalid turn journal")


def _span_kwargs(record: dict[str, Any], fallback_parent_id: str | None) -> dict[str, Any]:
    return _compact(
        {
            "id": record.get("id"),
            "parent_id": record.get("parentId", fallback_parent_id),
            "name": record.get("name") or "span",
            "type": record.get("type") or "span",
            "input": record.get("input"),
            "output": record.get("output"),
            "metadata": record.get("metadata"),
            "attributes": record.get("attributes"),
            "started_at": record.get("startedAt"),
            "ended_at": record.get("endedAt"),
            "duration_ms": record.get("durationMs"),
            "status": record.get("status"),
            "error": record.get("error"),
            "model": record.get("model"),
            "tool_name": record.get("toolName"),
            "usage": record.get("usage"),
            "user_facing_message": record.get("userFacingMessage"),
            "llm_provider": record.get("llmProvider"),
            "llm_model_name": record.get("llmModelName"),
            "llm_input_messages": record.get("llmInputMessages"),
            "llm_output_messages": record.get("llmOutputMessages"),
            "llm_invocation_parameters": record.get("llmInvocationParameters"),
            "llm_tools": record.get("llmTools"),
        }
    )


_START_SPAN_KEYS = {
    "name",
    "input",
    "metadata",
    "attributes",
    "id",
    "parent_id",
    "started_at",
}
_START_GENERATION_KEYS = _START_SPAN_KEYS | {
    "model",
    "llm_provider",
    "llm_invocation_parameters",
    "llm_input_messages",
    "llm_tools",
}
_START_TOOL_KEYS = _START_SPAN_KEYS | {"tool_name", "user_facing_message"}
_END_KEYS = {
    "output",
    "duration_ms",
    "status",
    "error",
    "ended_at",
    "metadata",
    "attributes",
    "model",
    "tool_name",
    "llm_provider",
    "llm_invocation_parameters",
    "llm_output_messages",
    "usage",
    "input_tokens",
    "output_tokens",
}


def _pick(kwargs: dict[str, Any], allowed: set[str]) -> dict[str, Any]:
    return {key: value for key, value in kwargs.items() if key in allowed}


def _start_from_record(context: TraceContext, kwargs: dict[str, Any]) -> SpanHandle:
    span_type = kwargs.get("type") or "span"
    if span_type == "generation":
        return context.start_generation(**_pick(kwargs, _START_GENERATION_KEYS))
    if span_type == "tool":
        return context.start_tool(**_pick(kwargs, _START_TOOL_KEYS))
    return context.start_span(**_pick(kwargs, _START_SPAN_KEYS))


def _record_from_record(context: TraceContext, kwargs: dict[str, Any]) -> None:
    context.span(**kwargs)


def apply_turn_journal(context: TraceContext, value: Any) -> None:
    token, records = _parse_journal(value)
    fallback_parent = token.get("parentSpanId") if token else None
    for record in records:
        if not isinstance(record, dict) or not record.get("id") or not record.get("op"):
            continue
        kwargs = _span_kwargs(record, fallback_parent)
        op = record["op"]
        span_id = record["id"]
        if op == "start":
            if context.has_span(span_id):
                continue
            _start_from_record(context, dict(kwargs))
            continue
        if op == "end":
            handle = context.span_handle(span_id)
            if handle is not None:
                handle.end(
                    **_pick(
                        _compact(
                            {
                                "output": record.get("output"),
                                "metadata": record.get("metadata"),
                                "attributes": record.get("attributes"),
                                "ended_at": record.get("endedAt"),
                                "duration_ms": record.get("durationMs"),
                                "status": record.get("status"),
                                "error": record.get("error"),
                                "model": record.get("model"),
                                "tool_name": record.get("toolName"),
                                "usage": record.get("usage"),
                                "llm_provider": record.get("llmProvider"),
                                "llm_output_messages": record.get("llmOutputMessages"),
                                "llm_invocation_parameters": record.get(
                                    "llmInvocationParameters"
                                ),
                            }
                        ),
                        _END_KEYS,
                    )
                )
                continue
            if context.has_span(span_id):
                continue
            _record_from_record(context, kwargs)
            continue
        if context.has_span(span_id):
            continue
        _record_from_record(context, kwargs)


def assemble_turn(
    token: str | dict[str, Any],
    journal: Any = None,
    *,
    name: str | None = None,
    input: Any = None,
    output: Any = None,
    metadata: dict[str, Any] | None = None,
    thread_id: str | None = None,
    user_id: str | None = None,
    duration_ms: int | None = None,
) -> tuple[TraceContext, datetime]:
    parsed = parse_turn_context_token(token)
    started_at = _datetime_or_now(parsed["startedAt"])
    context = TraceContext(
        id=parsed["traceId"],
        name=name or parsed.get("name") or "trace",
        input=input,
        metadata=metadata,
        thread_id=thread_id or parsed.get("threadId"),
        user_id=user_id or parsed.get("userId"),
        duration_ms=duration_ms,
    )
    if output is not None:
        context.output(output)
    if journal is not None:
        apply_turn_journal(context, journal)
    return context, started_at


def _record_fields(
    *,
    op: str,
    span_type: str,
    fallback_parent_id: str | None,
    **kwargs: Any,
) -> dict[str, Any]:
    started = kwargs.get("started_at")
    ended = kwargs.get("ended_at")
    record = _compact(
        {
            "op": op,
            "id": kwargs.get("id") or str(uuid.uuid4()),
            "parentId": kwargs.get("parent_id", fallback_parent_id),
            "name": kwargs.get("name"),
            "type": span_type,
            "input": kwargs.get("input"),
            "output": kwargs.get("output"),
            "metadata": kwargs.get("metadata"),
            "attributes": kwargs.get("attributes"),
            "startedAt": _iso(started) if started is not None else None,
            "endedAt": _iso(ended) if ended is not None else None,
            "durationMs": kwargs.get("duration_ms"),
            "status": kwargs.get("status"),
            "error": kwargs.get("error"),
            "model": kwargs.get("model"),
            "toolName": kwargs.get("tool_name"),
            "usage": kwargs.get("usage"),
            "userFacingMessage": kwargs.get("user_facing_message"),
            "llmProvider": kwargs.get("llm_provider"),
            "llmModelName": kwargs.get("llm_model_name"),
            "llmInputMessages": kwargs.get("llm_input_messages"),
            "llmOutputMessages": kwargs.get("llm_output_messages"),
            "llmInvocationParameters": kwargs.get("llm_invocation_parameters"),
            "llmTools": kwargs.get("llm_tools"),
        }
    )
    return {key: record[key] for key in ("op", *_RECORD_KEYS) if key in record}


class AttachedSpanHandle:
    def __init__(self, recorder: AttachedTurn, record: dict[str, Any]) -> None:
        self._recorder = recorder
        self._record = record
        self.id = record["id"]
        self._ended = record.get("op") != "start"

    def end(self, **kwargs: Any) -> None:
        if self._ended:
            return
        self._ended = True
        fields = {
            "id": self.id,
            "name": self._record.get("name"),
            "parent_id": self._record.get("parentId"),
            "input": self._record.get("input"),
            "model": self._record.get("model"),
            "tool_name": self._record.get("toolName"),
            **kwargs,
        }
        if "ended_at" not in kwargs:
            fields["ended_at"] = _now()
        self._recorder.append(
            _record_fields(
                op="end",
                span_type=self._record.get("type") or "span",
                fallback_parent_id=self._record.get("parentId"),
                **fields,
            )
        )

    def start_span(self, **kwargs: Any) -> AttachedSpanHandle:
        kwargs.setdefault("parent_id", self.id)
        return self._recorder.start_span(**kwargs)

    def start_generation(self, **kwargs: Any) -> AttachedSpanHandle:
        kwargs.setdefault("parent_id", self.id)
        return self._recorder.start_generation(**kwargs)

    def start_tool(self, **kwargs: Any) -> AttachedSpanHandle:
        kwargs.setdefault("parent_id", self.id)
        return self._recorder.start_tool(**kwargs)

    def record_span(self, **kwargs: Any) -> None:
        kwargs.setdefault("parent_id", self.id)
        self._recorder.record_span(**kwargs)

    def record_generation(self, **kwargs: Any) -> None:
        kwargs.setdefault("parent_id", self.id)
        self._recorder.record_generation(**kwargs)

    def record_tool(self, **kwargs: Any) -> None:
        kwargs.setdefault("parent_id", self.id)
        self._recorder.record_tool(**kwargs)


class AttachedTurn:
    def __init__(self, token: str | dict[str, Any]) -> None:
        self.token = parse_turn_context_token(token)
        self._records: list[dict[str, Any]] = []

    @property
    def id(self) -> str:
        return self.token["traceId"]

    def append(self, record: dict[str, Any]) -> None:
        self._records.append(record)

    def records(self) -> dict[str, Any]:
        return {
            "version": TURN_JOURNAL_VERSION,
            "token": self.token,
            "records": list(self._records),
        }

    def start_span(self, **kwargs: Any) -> AttachedSpanHandle:
        return self._start("span", kwargs)

    def start_generation(self, **kwargs: Any) -> AttachedSpanHandle:
        return self._start("generation", kwargs)

    def start_tool(self, **kwargs: Any) -> AttachedSpanHandle:
        return self._start("tool", kwargs)

    def record_span(self, **kwargs: Any) -> None:
        self._record("span", kwargs)

    def record_generation(self, **kwargs: Any) -> None:
        self._record("generation", kwargs)

    def record_tool(self, **kwargs: Any) -> None:
        self._record("tool", kwargs)

    def _start(self, span_type: str, kwargs: dict[str, Any]) -> AttachedSpanHandle:
        kwargs = dict(kwargs)
        kwargs.setdefault("id", str(uuid.uuid4()))
        kwargs.setdefault("started_at", _now())
        kwargs.setdefault("parent_id", self.token.get("parentSpanId"))
        record = _record_fields(
            op="start",
            span_type=span_type,
            fallback_parent_id=self.token.get("parentSpanId"),
            **kwargs,
        )
        self.append(record)
        return AttachedSpanHandle(self, record)

    def _record(self, span_type: str, kwargs: dict[str, Any]) -> None:
        kwargs = dict(kwargs)
        kwargs.setdefault("id", str(uuid.uuid4()))
        now = _now()
        kwargs.setdefault("started_at", now)
        if "ended_at" not in kwargs:
            kwargs["ended_at"] = now
        kwargs.setdefault("parent_id", self.token.get("parentSpanId"))
        self.append(
            _record_fields(
                op="record",
                span_type=span_type,
                fallback_parent_id=self.token.get("parentSpanId"),
                **kwargs,
            )
        )


class TurnHandle(TraceHandle):
    def export(self, parent_span_id: str | None = None) -> dict[str, Any]:
        return _compact(
            {
                "version": TURN_CONTEXT_VERSION,
                "traceId": self.id,
                "parentSpanId": parent_span_id,
                "threadId": self.thread_id,
                "userId": self.user_id,
                "startedAt": _iso(self.started_at),
                "name": self.name,
            }
        )

    def apply(self, journal: Any) -> None:
        apply_turn_journal(self, journal)


def start_turn(lemma: Lemma, **kwargs: Any) -> TurnHandle:
    return TurnHandle(lemma, **kwargs)


def attach_turn(token: str | dict[str, Any]) -> AttachedTurn:
    return AttachedTurn(token)

File: packages/ts/tracing/src/turn.ts
import {
  Lemma,
  SpanHandle,
  TraceContext,
  TraceHandle,
  type GenerationOptions,
  type SpanOptions,
  type SpanType,
  type TokenUsage,
  type ToolOptions,
  type TraceEndOptions,
  type TraceOptions,
} from "./client";

export const TURN_CONTEXT_VERSION = 1 as const;
export const TURN_JOURNAL_VERSION = 1 as const;

export type TurnContextToken = {
  version: typeof TURN_CONTEXT_VERSION;
  traceId: string;
  parentSpanId?: string | null;
  threadId?: string;
  userId?: string;
  startedAt: string;
  name?: string;
};

export type TurnJournalSpan = {
  id: string;
  parentId?: string | null;
  name?: string;
  type?: SpanType;
  input?: unknown;
  output?: unknown;
  metadata?: Record<string, unknown>;
  attributes?: Record<string, unknown>;
  startedAt?: string | null;
  endedAt?: string | null;
  durationMs?: number;
  status?: "OK" | "ERROR";
  error?: unknown;
  model?: string;
  toolName?: string;
  usage?: TokenUsage;
  userFacingMessage?: string;
  llmProvider?: string;
  llmModelName?: string;
  llmInputMessages?: unknown[];
  llmOutputMessages?: unknown[];
  llmInvocationParameters?: unknown;
  llmTools?: unknown;
};

export type TurnJournalRecord = TurnJournalSpan & {
  op: "start" | "end" | "record";
};

export type TurnJournal = {
  version: typeof TURN_JOURNAL_VERSION;
  token: TurnContextToken;
  records: TurnJournalRecord[];
};

export type TurnJournalInput =
  | TurnJournal
  | TurnJournalRecord
  | TurnJournalRecord[]
  | string;

function compact<T extends Record<string, unknown>>(value: T): T {
  return Object.fromEntries(
    Object.entries(value).filter(([, entry]) => entry !== undefined),
  ) as T;
}

function isoNow(): string {
  return new Date().toISOString();
}

function asIso(value: Date | string | null | undefined): string | null | undefined {
  if (value == null) return value;
  return value instanceof Date ? value.toISOString() : value;
}

export function parseTurnContextToken(
  value: string | TurnContextToken,
): TurnContextToken {
  const token =
    typeof value === "string"
      ? (JSON.parse(value) as TurnContextToken)
      : value;
  if (
    !token ||
    typeof token !== "object" ||
    token.version !== TURN_CONTEXT_VERSION ||
    typeof token.traceId !== "string" ||
    token.traceId.length === 0 ||
    typeof token.startedAt !== "string" ||
    token.startedAt.length === 0
  ) {
    throw new Error("@uselemma/tracing: invalid turn context token");
  }
  return token;
}

function parseJournal(input: TurnJournalInput): {
  token?: TurnContextToken;
  records: TurnJournalRecord[];
} {
  const value =
    typeof input === "string" ? (JSON.parse(input) as unknown) : input;
  if (Array.isArray(value)) {
    return { records: value };
  }
  if (value && typeof value === "object" && "op" in value) {
    return { records: [value as TurnJournalRecord] };
  }
  if (
    value &&
    typeof value === "object" &&
    "records" in value &&
    Array.isArray((value as TurnJournal).records)
  ) {
    const journal = value as TurnJournal;
    if (
      journal.version !== undefined &&
      journal.version !== TURN_JOURNAL_VERSION
    ) {
      throw new Error(
        `@uselemma/tracing: unsupported turn journal version ${String(journal.version)}`,
      );
    }
    return { token: journal.token, records: journal.records };
  }
  throw new Error("@uselemma/tracing: invalid turn journal");
}

function toSpanOptions(
  record: TurnJournalRecord,
  fallbackParentId?: string | null,
): SpanOptions {
  return compact({
    id: record.id,
    parentId: record.parentId ?? fallbackParentId,
    name: record.name ?? "span",
    type: record.type,
    input: record.input,
    output: record.output,
    metadata: record.metadata,
    attributes: record.attributes,
    startedAt: record.startedAt,
    endedAt: record.endedAt,
    durationMs: record.durationMs,
    status: record.status,
    error: record.error,
    model: record.model,
    toolName: record.toolName,
    usage: record.usage,
    userFacingMessage: record.userFacingMessage,
    llmProvider: record.llmProvider,
    llmModelName: record.llmModelName,
    llmInputMessages: record.llmInputMessages,
    llmOutputMessages: record.llmOutputMessages,
    llmInvocationParameters: record.llmInvocationParameters,
    llmTools: record.llmTools,
  });
}

function startFromRecord(
  context: TraceContext,
  options: SpanOptions,
): SpanHandle {
  const type = options.type ?? "span";
  if (type === "generation") {
    return context.startGeneration(options);
  }
  if (type === "tool") {
    return context.startTool(options);
  }
  return context.startSpan(options);
}

function recordFromRecord(context: TraceContext, options: SpanOptions) {
  const type = options.type ?? "span";
  if (type === "generation") {
    context.recordGeneration(options);
    return;
  }
  if (type === "tool") {
    context.recordTool(options);
    return;
  }
  context.recordSpan(options);
}

/**
 * Hydrate a coordinator {@link TraceContext} from a child-process journal.
 * Re-applying the same records is a no-op for already-present span ids.
 */
export function applyTurnJournal(
  context: TraceContext,
  input: TurnJournalInput,
): void {
  const { token, records } = parseJournal(input);
  const fallbackParentId = token?.parentSpanId ?? null;
  for (const record of records) {
    if (!record?.id || !record.op) continue;
    const options = toSpanOptions(record, fallbackParentId);
    if (record.op === "start") {
      if (context.hasSpan(record.id)) continue;
      startFromRecord(context, { ...options, endedAt: undefined });
      continue;
    }
    if (record.op === "end") {
      const handle = context.spanHandle(record.id);
      if (handle) {
        const endOptions: Omit<SpanOptions, "id" | "name" | "type" | "startedAt"> & {
          userFacingMessage?: string;
        } = {
          output: record.output,
          metadata: record.metadata,
          attributes: record.attributes,
          endedAt: record.endedAt ?? undefined,
          durationMs: record.durationMs,
          status: record.status,
          error: record.error,
          model: record.model,
          toolName: record.toolName,
          usage: record.usage,
          userFacingMessage: record.userFacingMessage,
          llmProvider: record.llmProvider,
          llmModelName: record.llmModelName,
          llmOutputMessages: record.llmOutputMessages,
          llmInvocationParameters: record.llmInvocationParameters,
          llmTools: record.llmTools,
        };
        handle.end(endOptions);
        continue;
      }
      if (context.hasSpan(record.id)) continue;
      recordFromRecord(context, options);
      continue;
    }
    if (context.hasSpan(record.id)) continue;
    recordFromRecord(context, options);
  }
}

/**
 * Build a {@link TraceContext} from a context token and optional journal,
 * ready for a single strict {@link Lemma.ingest} call.
 */
export function assembleTurn(
  token: string | TurnContextToken,
  journal?: TurnJournalInput,
  options: Omit<TraceOptions, "id"> = {},
): { context: TraceContext; startedAt: Date } {
  const parsed = parseTurnContextToken(token);
  const context = new TraceContext({
    id: parsed.traceId,
    name: options.name ?? parsed.name ?? "trace",
    input: options.input,
    output: options.output,
    metadata: options.metadata,
    threadId: options.threadId ?? parsed.threadId,
    userId: options.userId ?? parsed.userId,
    durationMs: options.durationMs,
    startedAt: options.startedAt ?? parsed.startedAt,
  });
  if (journal !== undefined) applyTurnJournal(context, journal);
  return { context, startedAt: new Date(parsed.startedAt) };
}

function spanFields(
  options: SpanOptions & ToolOptions & { type?: SpanType },
  fallbackParentId?: string | null,
): TurnJournalSpan {
  const toolOptions = options as ToolOptions;
  return compact({
    id: options.id ?? crypto.randomUUID(),
    parentId:
      options.parentId ?? options.parentSpanId ?? fallbackParentId ?? null,
    name: options.name,
    type: options.type,
    input: options.input,
    output: options.output,
    metadata: options.metadata,
    attributes: options.attributes,
    startedAt: asIso(options.startedAt),
    endedAt: asIso(options.endedAt),
    durationMs: options.durationMs,
    status: options.status,
    error: options.error,
    model: options.model,
    toolName: toolOptions.toolName,
    usage: options.usage,
    userFacingMessage: toolOptions.userFacingMessage,
    llmProvider: options.llmProvider,
    llmModelName: options.llmModelName,
    llmInputMessages: options.llmInputMessages,
    llmOutputMessages: options.llmOutputMessages,
    llmInvocationParameters: options.llmInvocationParameters,
    llmTools: options.llmTools,
  });
}

export class AttachedSpanHandle {
  readonly id: string;

  constructor(
    private readonly recorder: AttachedTurn,
    private readonly fields: TurnJournalSpan,
    private ended = false,
  ) {
    this.id = fields.id;
  }

  end(
    options: Omit<SpanOptions, "id" | "name" | "type" | "startedAt"> = {},
  ) {
    if (this.ended) return;
    this.ended = true;
    this.recorder.append({
      op: "end",
      ...spanFields(
        {
          ...this.fields,
          ...options,
          id: this.id,
          name: this.fields.name ?? "span",
        },
        this.fields.parentId,
      ),
      id: this.id,
      startedAt: this.fields.startedAt,
    });
  }

  startSpan(name: string): AttachedSpanHandle;
  startSpan(options: Omit<SpanOptions, "endedAt">): AttachedSpanHandle;
  startSpan(
    options: string | Omit<SpanOptions, "endedAt">,
  ): AttachedSpanHandle {
    const spanOptions =
      typeof options === "string" ? { name: options } : options;
    return this.recorder.startSpan({
      ...spanOptions,
      parentId: spanOptions.parentId ?? this.id,
    });
  }

  startGeneration(name: string): AttachedSpanHandle;
  startGeneration(
    options: Omit<GenerationOptions, "endedAt" | "type">,
  ): AttachedSpanHandle;
  startGeneration(
    options: string | Omit<GenerationOptions, "endedAt" | "type">,
  ): AttachedSpanHandle {
    const generationOptions =
      typeof options === "string" ? { name: options } : options;
    return this.recorder.startGeneration({
      ...generationOptions,
      parentId: generationOptions.parentId ?? this.id,
    });
  }

  startTool(name: string): AttachedSpanHandle;
  startTool(options: Omit<ToolOptions, "endedAt" | "type">): AttachedSpanHandle;
  startTool(
    options: string | Omit<ToolOptions, "endedAt" | "type">,
  ): AttachedSpanHandle {
    const toolOptions =
      typeof options === "string" ? { name: options } : options;
    return this.recorder.startTool({
      ...toolOptions,
      parentId: toolOptions.parentId ?? this.id,
    });
  }

  recordSpan(name: string): AttachedSpanHandle;
  recordSpan(options: SpanOptions): AttachedSpanHandle;
  recordSpan(options: string | SpanOptions): AttachedSpanHandle {
    const spanOptions =
      typeof options === "string" ? { name: options } : options;
    return this.recorder.recordSpan({
      ...spanOptions,
      parentId: spanOptions.parentId ?? this.id,
    });
  }

  recordGeneration(options: string | GenerationOptions) {
    const generationOptions =
      typeof options === "string" ? { name: options } : options;
    this.recorder.recordGeneration({
      ...generationOptions,
      parentId: generationOptions.parentId ?? this.id,
    });
  }

  recordTool(options: string | ToolOptions) {
    const toolOptions =
      typeof options === "string" ? { name: options } : options;
    this.recorder.recordTool({
      ...toolOptions,
      parentId: toolOptions.parentId ?? this.id,
    });
  }
}

/** Local recorder for a worker/sandbox process. Never calls Lemma. */
export class AttachedTurn {
  readonly token: TurnContextToken;
  private readonly journal: TurnJournalRecord[] = [];

  constructor(token: string | TurnContextToken) {
    this.token = parseTurnContextToken(token);
  }

  get id() {
    return this.token.traceId;
  }

  append(record: TurnJournalRecord) {
    this.journal.push(record);
  }

  records(): TurnJournal {
    return {
      version: TURN_JOURNAL_VERSION,
      token: this.token,
      records: this.journal,
    };
  }

  startSpan(name: string): AttachedSpanHandle;
  startSpan(options: Omit<SpanOptions, "endedAt">): AttachedSpanHandle;
  startSpan(
    options: string | Omit<SpanOptions, "endedAt">,
  ): AttachedSpanHandle {
    const spanOptions =
      typeof options === "string" ? { name: options } : options;
    return this.start("span", spanOptions);
  }

  startGeneration(name: string): AttachedSpanHandle;
  startGeneration(
    options: Omit<GenerationOptions, "endedAt" | "type">,
  ): AttachedSpanHandle;
  startGeneration(
    options: string | Omit<GenerationOptions, "endedAt" | "type">,
  ): AttachedSpanHandle {
    const generationOptions =
      typeof options === "string" ? { name: options } : options;
    return this.start("generation", generationOptions);
  }

  startTool(name: string): AttachedSpanHandle;
  startTool(options: Omit<ToolOptions, "endedAt" | "type">): AttachedSpanHandle;
  startTool(
    options: string | Omit<ToolOptions, "endedAt" | "type">,
  ): AttachedSpanHandle {
    const toolOptions =
      typeof options === "string" ? { name: options } : options;
    return this.start("tool", toolOptions);
  }

  recordSpan(name: string): AttachedSpanHandle;
  recordSpan(options: SpanOptions): AttachedSpanHandle;
  recordSpan(options: string | SpanOptions): AttachedSpanHandle {
    const spanOptions =
      typeof options === "string" ? { name: options } : options;
    return this.record("span", spanOptions);
  }

  recordGeneration(options: string | GenerationOptions) {
    const generationOptions =
      typeof options === "string" ? { name: options } : options;
    this.record("generation", generationOptions);
  }

  recordTool(options: string | ToolOptions) {
    const toolOptions =
      typeof options === "string" ? { name: options } : options;
    this.record("tool", toolOptions);
  }

  private start(
    type: SpanType,
    options: Omit<SpanOptions, "endedAt">,
  ): AttachedSpanHandle {
    const fields = spanFields(
      {
        ...options,
        type,
        id: options.id ?? crypto.randomUUID(),
        startedAt: options.startedAt ?? isoNow(),
      },
      this.token.parentSpanId,
    );
    this.append({ op: "start", ...fields, type });
    return new AttachedSpanHandle(this, { ...fields, type });
  }

  private record(
    type: SpanType,
    options: SpanOptions,
  ): AttachedSpanHandle {
    const now = isoNow();
    const fields = spanFields(
      {
        ...options,
        type,
        id: options.id ?? crypto.randomUUID(),
        startedAt: options.startedAt ?? now,
        endedAt: options.endedAt === undefined ? now : options.endedAt,
      },
      this.token.parentSpanId,
    );
    this.append({ op: "record", ...fields, type });
    return new AttachedSpanHandle(this, { ...fields, type }, true);
  }
}

export class TurnHandle extends TraceHandle {
  export(options: { parentSpanId?: string | null } = {}): TurnContextToken {
    const identity = this.identity();
    return compact({
      version: TURN_CONTEXT_VERSION,
      traceId: this.id,
      parentSpanId: options.parentSpanId ?? null,
      threadId: identity.threadId,
      userId: identity.userId,
      startedAt: this.startedAt.toISOString(),
      name: identity.name,
    });
  }

  apply(input: TurnJournalInput) {
    applyTurnJournal(this, input);
  }
}

export function startTurn(
  lemma: Lemma,
  options: TraceOptions = {},
): TurnHandle {
  return new TurnHandle(
    options,
    async (trace, startedAt, endedAt) => {
      await lemma.ingest(trace, { startedAt, endedAt });
    },
  );
}

export function attachTurn(
  token: string | TurnContextToken,
): AttachedTurn {
  return new AttachedTurn(token);
}
